src/upf/src/upf.cpp

```diff
@@ -874,10 +866,342 @@ static bool isolate_port(utl::Logger* logger,
                           output_m,
                           target_group);
 
-    for (auto&& external_iterm : external_iterms) {
-      external_iterm->disconnect();
-      external_iterm->connect(net_out);
+    target_iterm->disconnect();
+    target_iterm->connect(net_out);
+  }
+
+  return true;
+}
+
+// Returns all connected iterms that are not in the same power domain
+static std::vector<std::pair<odb::dbITerm*, odb::dbPowerDomain*>>
+get_connected_terms(odb::dbBlock* block, odb::dbITerm* iterm)
+{
+  std::vector<std::pair<odb::dbITerm*, odb::dbPowerDomain*>> external_iterms;
+  auto net = iterm->getNet();
+
+  if (!net) {
+    return external_iterms;
+  }
+
+  auto connectedIterms = net->getITerms();
+
+  if (connectedIterms.size() < 2) {
+    return external_iterms;
+  }
+
+  for (auto&& connectedIterm : connectedIterms) {
+    if (connectedIterm == iterm) {
+      continue;
     }
+
+    auto connectedInst = connectedIterm->getInst();
+    // TODO: if connectedInst is isolation cell or level shifting cell
+    // then ignore it
+
+    if (!connectedInst->getGroup()
+        || connectedInst->getGroup() != iterm->getInst()->getGroup()) {
+      odb::dbPowerDomain* connectedDomain = nullptr;
+      if (connectedInst->getGroup()) {
+        connectedDomain
+            = block->findPowerDomain(connectedInst->getGroup()->getName());
```

Comment:
It would be nicer if you could do connectedDomain = connectedInst->getPowerDomain();